### PR TITLE
Added IAM Role Authentication using Web Identity Token

### PR DESCRIPTION
**Pull request by maliankurh**

### CODE REVIEW

**Comment by dekobon** [comment]
Hi @maliankurh, are you willing / interested on further refining this PR? If not, my apologies for saying "do a PR" because I just realized that I can't contribute to the branch that you did the PR on.

File: common/etc/nginx/include/s3gateway.js [hunk 1]
```diff
@@ -994,6 +1003,38 @@ async function _fetchEC2RoleCredentials() {
...
+ * @private
+ */
+async function _fetchWebIdentityCredentials(r) {
+    var arn   = process.env['AWS_ROLE_ARN'];
+    var name  = process.env['HOSTNAME'] || 'nginx-s3-gateway';
+    var token = fs.readFileSync(process.env['AWS_WEB_IDENTITY_TOKEN_FILE']);
+    
+    var params = "Version=2011-06-15&Action=AssumeRoleWithWebIdentity&RoleArn=" + arn + "&RoleSessionName=" + name + "&WebIdentityToken=" + token;
+
+    var response = await ngx.fetch("https://sts.us-east-1.amazonaws.com?" + params, {
```
Comment (mvanholsteijn) [inline common/etc/nginx/include/s3gateway.js:1020]: _By default, the AWS Security Token Service (AWS STS) is available as a global service, and all STS requests go to a single endpoint at https://sts.amazonaws.com/. AWS recommends using Regional STS endpoints to reduce latency, build in redundancy, and increase session token validity._

https://docs.aws.amazon.com/general/latest/gr/sts.html#sts_region

I would recommend adding an environment AWS_STS_ENDPOINT to configure the desired endpoint, while defaulting to sts.amazonaws.com.

**Comment by maliankur** [comment]
> Hi @maliankurh, are you willing / interested on further refining this PR? If not, my apologies for saying "do a PR" because I just realized that I can't contribute to the branch that you did the PR on.

Hello @dekobon, Sure Let me know what changes you want me to make for the PR.

File: common/etc/nginx/include/s3gateway.js (hunk 1 above)
Comment (maliankur) [reply common/etc/nginx/include/s3gateway.js:1020]: Hello @mvanholsteijn, sure I will make the changes accordingly.

**Comment by mvanholsteijn** [comment]
@dekobon can you specify what changes you need, so that this PR can go through?

File: common/etc/nginx/include/s3gateway.js [hunk 2]
```diff
@@ -910,6 +910,15 @@ async function fetchCredentials(r) {
             r.return(500);
             return;
         }
+    }
+    else if(process.env['AWS_WEB_IDENTITY_TOKEN_FILE']) {
```
Comment (dekobon) [inline common/etc/nginx/include/s3gateway.js:914]: Please document this environment variable in the documentation.

File: common/etc/nginx/include/s3gateway.js (hunk 1 above)
Comment (dekobon) [reply common/etc/nginx/include/s3gateway.js:1020]: I agree with the above assessment.

File: common/etc/nginx/include/s3gateway.js [hunk 3]
```diff
@@ -994,6 +1003,38 @@ async function _fetchEC2RoleCredentials() {
...
+/**
+ * Get the credentials by assuming calling AssumeRoleWithWebIdentity with the environment variable
+ * values ROLE_ARN, AWS_WEB_IDENTITY_TOKEN_FILE and HOSTNAME
+ *
+ * @returns {Promise<{accessKeyId: (string), secretAccessKey: (string), sessionToken: (string), expiration: (string)}>}
+ * @private
+ */
+async function _fetchWebIdentityCredentials(r) {
+    var arn   = process.env['AWS_ROLE_ARN'];
+    var name  = process.env['HOSTNAME'] || 'nginx-s3-gateway';
```
Comment (dekobon) [inline common/etc/nginx/include/s3gateway.js:1015]: Please document all new environment variables.

**Review by dekobon — CHANGES_REQUESTED** [review]
In addition to the requested changes, please add:
* Automated tests that validate the configuration (if possible)
* Use case documentation that describes what you would accomplish and how you would configure the gateway

Also, please test the other host credential based access configurations in order to see if this change introduces a regression.

Thank you so much for your contribution!

**Comment by maliankurh** [comment]
@dekobon, I have updated the Documentation and added the steps for the STS to work. Please check and let me know if you want me to update anything else.

@mvanholsteijn, I have added Env Variable for STS Endpoint and set the default value as well. Please check

**Comment by dekobon** [comment]
It seems that the automated test run is failing. Can you look into why the tests are failing: https://github.com/nginxinc/nginx-s3-gateway/runs/8277852689?check_suite_focus=true

**Comment by maliankurh** [comment]
Yes I looked into it and the latest commit should fix the issue.

**Comment by dekobon** [comment]
Awesome, let's rebase the latest changes from master into this branch and I will try to do some testing with the branch myself. Let's aim for a merge next week.

If anyone else who is following the PR would like to do their own testing, that would be much appreciated.

**Comment by dekobon** [comment]
I want to set expectations for everyone clearly. I've been unable to do the testing that I wanted to this week. I'll aim to be working on it starting on 9/19. My apologies for the delay folks.

**Comment by dm3ch** [comment]
I just tested this PR in my own EKS, cause I have similar case.

What I've found:
- It's not working for me without https://github.com/nginxinc/nginx-s3-gateway/pull/52/files#diff-3077c17ca3f4dd131f26585ea1026611c248ff1b3268e70c17a33b35607e2d58R82
- I'm not sure why (maybe cause of version upgrade in master or different credentials) but for me `PROVIDE_INDEX_PAGE` stoped working

**Comment by dekobon** [comment]
I just tried to run: `./test.sh oss` using this branch and the test script returned the following failure:
```
...
▶ Building Docker Compose environment
[+] Running 3/3
 ⠿ Container ngt-minio-1             Created                                                                       0.8s
 ⠿ Container ngt-minio-client-1      Created                                                                       0.6s
 ⠿ Container ngt-nginx-s3-gateway-1  Created                                                                       0.8s
▶ Adding test data to container
Copying contents of /home/elijah/Development/F5/nginx-s3-gateway/test/data to Docker container ngt_minio_1:/
Error: No such container:path: ngt_minio_1:/
Error running tests - outputting container logs
```

I'm going to investigate the cause and update this discussion.

**Comment by dekobon** [comment]
After going through the differences between this branch and master, it seems like it is a number of commits behind master. I think it would be best to create a new branch under the main project with the contributions in this PR, merge in the changes from master, and then try to pull in as many volunteers into testing it as possible.

**Comment by dekobon** [comment]
I've created a new branch: https://github.com/nginxinc/nginx-s3-gateway/tree/issue/40
And a new PR for integrating that branch into master: https://github.com/nginxinc/nginx-s3-gateway/pull/54

If anyone has fixes or improvements, please do a PR against the [issue/40](https://github.com/nginxinc/nginx-s3-gateway/tree/issue/40) branch.

In this new branch, I've merged in the latest changes from master. After doing that, the automated tests now pass.

I'm going to close this PR and have us move our development to the other PR. I hope this isn't too confusing. I'm eager to help and see this feature added.

**Pull request closed without merging**